src/planner/match/Expand.cpp

```diff
@@ -110,34 +112,58 @@ Status Expand::expandSteps(const NodeInfo& node,
                                                               initialExpr_.release(),
                                                               inputVar_,
                                                               subplan));
-        // If maxHop > 0, the result of 0 step will be passed to next plan node
-        if (maxHop > 0) {
-            subplan.root = passThrough(matchCtx_->qctx, subplan.root);
-        }
     } else {  // Case 1 to n steps
         startIndex = 1;
         // Expand first step from src
         NG_RETURN_IF_ERROR(expandStep(edge, dependency_, inputVar_, node.filter, &subplan));
-        // Manualy create a passThrough node for the first step
-        // Rest steps will be passed through in collectData()
-        subplan.root = passThrough(matchCtx_->qctx, subplan.root);
     }
-
-    PlanNode* passThrough = subplan.root;
-    for (; startIndex < maxHop; ++startIndex) {
-        SubPlan curr;
-        NG_RETURN_IF_ERROR(
-            expandStep(edge, passThrough, passThrough->outputVar(), nullptr, &curr));
-        auto rNode = subplan.root;
-        DCHECK(rNode->kind() == PNKind::kUnion || rNode->kind() == PNKind::kPassThrough);
-        NG_RETURN_IF_ERROR(collectData(passThrough, curr.root, rNode, &passThrough, &subplan));
+    // No need to further expand if maxHop is the start Index
+    if (maxHop == startIndex) {
+        plan->root = subplan.root;
+        return Status::OK();
     }
+    // Result of first step expansion
+    PlanNode* firstStep = subplan.root;
+
+    // Build Start node from first step
+    SubPlan loopBodyPlan;
+    PlanNode* startNode = StartNode::make(matchCtx_->qctx);
+    startNode->setOutputVar(firstStep->outputVar());
+    startNode->setColNames(firstStep->colNames());
+    loopBodyPlan.tail = startNode;
+    loopBodyPlan.root = startNode;
+    PlanNode* passThrough = loopBodyPlan.root;
```

Comment:
delete this useless variable `passThrough` and use startNode directly in following function procedure.